packages/report-server/src/__tests__/reportBuilder/transform/insert.test.ts

```diff
@@ -0,0 +1,124 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { SINGLE_ANALYTIC, MULTIPLE_ANALYTICS } from './transform.fixtures';
+import { buildTransform } from '../../../reportBuilder/transform';
+
+describe('insert', () => {
+  // SAME AS SELECT FUNCTIONALITY
+  it('can insert row with basic values', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'number'": '1',
+        "'string'": "'Hi'",
+        "'boolean'": 'false',
+      },
+    ]);
+    expect(transform(SINGLE_ANALYTIC)).toEqual([...SINGLE_ANALYTIC, { number: 1, string: 'Hi', boolean: false }]);
+  });
+
+  it('can insert row with values from previous row', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'dataElementValue'": '$row.value',
+      },
+    ]);
+    expect(transform(SINGLE_ANALYTIC)).toEqual([...SINGLE_ANALYTIC, { dataElementValue: 4 }]);
+  });
+
+  it('can select a value from the row as a field name', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        '$row.dataElement': '$row.value',
+      },
+    ]);
+    expect(transform(SINGLE_ANALYTIC)).toEqual([...SINGLE_ANALYTIC, { BCD1: 4 }]);
+  });
+
+  it('can execute functions', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'period'": "periodToDisplayString($row.period, 'DAY')",
+      },
+    ]);
+    expect(transform(SINGLE_ANALYTIC)).toEqual([...SINGLE_ANALYTIC, { period: '1st Jan 2020' }]);
+  });
+
+  // SPECIFIC TO INSERT
+  it('can insert multiple rows', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'dataElementValue'": '$row.value',
+      },
+    ]);
+    expect(transform(MULTIPLE_ANALYTICS)).toEqual([
+      { period: '20200101', organisationUnit: 'TO', dataElement: 'BCD1', value: 4 },
+      { dataElementValue: 4 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { dataElementValue: 2 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
+      { dataElementValue: 5 },
+    ]);
+  });
+
+  it('can insert new rows before the relative row', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'dataElementValue'": '$row.value',
+        position: 'before',
+      },
+    ]);
+    expect(transform(MULTIPLE_ANALYTICS)).toEqual([
+      { dataElementValue: 4 },
+      { period: '20200101', organisationUnit: 'TO', dataElement: 'BCD1', value: 4 },
+      { dataElementValue: 2 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { dataElementValue: 5 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
```

Comment:
Apologies for jumping in on this review, I was super curious! 🙏 

It's looking really good, and I very much appreciate the tests!!

Wanted to note that I'm a bit surprised that the default behaviour is to insert a row for each existing row if no `where` is specified. 

I do see the logic, and it's consistent with our other existing transforms that basically treat having no `where` clause as saying 'match every row'. 

However I feel it's a bit unexpected, I guess I had assumed that by default we'd just insert a single row (at the end by default)? And we'd only do the 'insert per matching row' if a `where` clause is specified?

Curious to hear your thoughts on this?

Reply:
@rohan-bes I looked at doing that to start with but I didn't like how inconsistent it felt.

Things like what does `position: 'before'` mean if there's no `where` clause? What happens if you reference `$row` when there's no `where` clause?

It winds up sorta being two separate functions depending on whether you have a `where`, and that started feeling a little messy in code.

I agree that the default behaviour is perhaps a bit unexpected, I can certainly change it to work like you've mentioned, or perhaps I could split the functionality into single/multi insert.

What would you prefer?

Reply:
No I do see your point... I think I could be won over pretty easily to the current implementation if there were an easy way to insert a single row. Perhaps something like `where: $row.index = 1` (single row at start) and `where: $row.index = $table.length` (where table is some lookup I'd suggest that is just an array of all rows, single row at end).

I'm not suggesting we do this now btw, if the transform you've got here meets all your requirements then that's all we need 😄 Just wanna make sure we have some kind of plan if we feel we need a single row insert behaviour...

Reply:
Cool, I've added a couple of extra lookup, `$index`, `$previous` and `$table` just because they're all useful for random things we might wanna do with insertions

I also added `length(Array)` as function because there was some weirdness with the parsing of `Array.length` format in reports.

And I've added some extra unit tests showing the single row at start, single row at end.

Reply:
Ah awesome! Thanks for adding those lookups, I feel they'll definitely come in handy!